source/MRMesh/MRCylinderApproximator.h

```diff
@@ -0,0 +1,373 @@
+#include "MRMeshFwd.h"
+#include "MRCylinder3.h"
+#include "MRVector.h"
+#include "MRMatrix.h"
+#include <Eigen/Eigenvalues>
+#include "MRPch/MRTBB.h"
+#include "MRToFromEigen.h"
+#include "MRConstants.h"
+
+// https://www.geometrictools.com/Documentation/LeastSquaresFitting.pdf 
+
+namespace MR
+{
+template <typename T>
+class Cylinder3Approximation
+{
+
+private:
+
+    enum class CylinderFitterType
+    {
+        // The algorithm implimentation needs an initial approximation to refine the cylinder axis. 
+        // In this option, we sort through several possible options distributed over the hemisphere.
+        HemisphereSearchFit,
+
+        // In this case, we assume that there is an external estimate for the cylinder axis. 
+        // Therefore, we specify only the position that is given from the outside
+        SpecificAxisFit
+
+        // TODO for Meshes try to impliment specific algorithm from https://www.geometrictools.com/Documentation/LeastSquaresFitting.pdf
+        // TODO Also, an estimate of the cylinder axis can be obtained by the gravel component method or the like. But this requires additional. experiments.
+        // TODO for future try eigen vector covariance   https://www.geometrictools.com/Documentation/RobustEigenSymmetric3x3.pdf
+    };
+
+    CylinderFitterType fitter_ = CylinderFitterType::HemisphereSearchFit;
+
+    //  CylinderFitterType::SpecificAxisFit params
+    Eigen::Vector<T, 3> baseCylinderAxis_;
+
+    // CylinderFitterType::HemisphereSearchFit params
+    size_t thetaResolution_ = 0;
+    size_t phiResolution_ = 0;
+    bool isMultithread_ = true;
+
+    //Input data converted to Eigen format and normalized to the avgPoint position of all points in the cloud.
+    std::vector<Eigen::Vector<T, 3>> normalizedPoints_ = {};
+
+    // Precalculated values for speed up. 
+    // In https://www.geometrictools.com/Documentation/LeastSquaresFitting.pdf page 35-36: 
+    // Text below is a direct copy from pdf file: 
+    // The sample application that used equation (94) directly was really slow. 
+    // On an Intel� CoreTM i7-6700 CPU at 3.40 GHz, the single - threaded version for 10765 points required 129 seconds 
+    // and the multithreaded version using 8 hyperthreads required 22 seconds.The evaluation of G using the precomputed summations is much faster.
+    // The single - threaded version required 85 milliseconds and the multithreaded version using 8 hyperthreads required 22 milliseconds.
+
+    Eigen::Vector <T, 6>  precomputedMu_ = {};
```

Comment:
`= {}` does not initialize Eigen Vector with zeros, but only produce compilation errors. To fill with zeros use `.setZero()` method (or probably there is something else in Eigen)

Reply:
done